server/api/src/rsModAccessControl.cpp

```diff
@@ -1,68 +1,95 @@
-    if ( strcmp( newPath, newModAccessControlInp.path ) != 0 ) {
+    char newPath[MAX_NAME_LEN]{};
+    rstrcpy(newPath, newModAccessControlInp.path, MAX_NAME_LEN);
+    resolveLinkedPath(rsComm, newPath, &specCollCache, nullptr);
+    if (strcmp(newPath, newModAccessControlInp.path)) {
         newModAccessControlInp.path = newPath;
     }
 
-    status = getAndConnRcatHost(
-                 rsComm,
-                 MASTER_RCAT,
-                 ( const char* )newModAccessControlInp.path,
-                 &rodsServerHost );
-    if ( status < 0 ) {
-        return status;
-    }
+    try {
+        // If the zone name cannot be derived from the collection name in the input, the path cannot be valid.
+        const auto zone_name = fs::zone_name(fs::path{static_cast<char*>(newModAccessControlInp.path)});
+        if (!zone_name) {
+            return SYS_INVALID_FILE_PATH;
+        }
+
+        if (!isLocalZone(zone_name->data())) {
+            auto* host = ic::redirect_to_catalog_provider(*rsComm, static_cast<char*>(newModAccessControlInp.path));
+            return rcModAccessControl(host->conn, &newModAccessControlInp);
+        }
 
-    if ( rodsServerHost->localFlag == LOCAL_HOST ) {
-        std::string svc_role;
-        irods::error ret = get_catalog_service_role(svc_role);
-        if(!ret.ok()) {
-            irods::log(PASS(ret));
-            return ret.code();
+        const auto catalog_provider_host = ic::get_catalog_provider_host();
+
+        if (LOCAL_HOST == catalog_provider_host.localFlag) {
+            ic::throw_if_catalog_provider_service_role_is_invalid();
+            return _rsModAccessControl(rsComm, &newModAccessControlInp);
         }
 
-        if( irods::CFG_SERVICE_ROLE_PROVIDER == svc_role ) {
-            status = _rsModAccessControl( rsComm, &newModAccessControlInp );
-        } else if( irods::CFG_SERVICE_ROLE_CONSUMER == svc_role ) {
-            status = SYS_NO_RCAT_SERVER_ERR;
-        } else {
-            rodsLog(
-                LOG_ERROR,
-                "role not supported [%s]",
-                svc_role.c_str() );
-            status = SYS_SERVICE_ROLE_NOT_SUPPORTED;
+        // Some privileged clients may be affected by temporary elevated privileges. The redirect to the catalog service
+        // provider here will cause the temporary privileged status granted to this connection to be lost in the
+        // server-to-server connection. The compound resource temporarily elevates permissions for users with read-only
+        // access to data objects which need to be staged to cache. A client connection is made here using the service
+        // account rodsadmin credentials to modify the permissions and then is immediately disconnected. This is only
+        // done for connections with elevated privileges that are not rodsadmins.
+        if (irods::is_privileged_client(*rsComm) &&
+            ua::user_type::rodsadmin != *ua::server::type(*rsComm, ua::user{rsComm->clientUser.userName}))
```

Comment:
Not sure... I guess we can make sure by passing `rsComm->clientUser.rodsZone`, right?

I will also add a test - similar to the one I've already written - to the federation test suite to make sure things work as expected.

Reply:
Sure.

Reply:
So, it turns out that the existing solution does not fix the issue for remote users (observed a cache replica stuck in the intermediate status).

This is because `scoped_permissions` also does not consider zone name when setting the permissions on the data object. I have adjusted the `scoped_permissions` class to hold a qualified username which is passed to the `irods::experimental::filesystem::permissions` call, which parses and interprets qualified usernames.

We should also make sure the user is correct by passing the zone name here, here but that did not resolve the issue (the resolution is noted in the above paragraph).

Reply:
Done